aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -171,38 +172,190 @@ Tensor ger(const Tensor& self, const Tensor& vec2) {
   return result;
 }
 
-Tensor addbmm_cpu(const Tensor& self, const Tensor& batch1, const Tensor& batch2, Scalar beta, Scalar alpha) {
-  Tensor b_self;
-  std::tie(b_self) = expand_size(self, {batch1.size(1), batch2.size(2)}, "addbmm");
-  return legacy::cpu::_th_addbmm(b_self, batch1, batch2, beta, alpha);
+static void addmm_impl_cpu_(
+    Tensor &result, const Tensor &self, Tensor m1, Tensor m2, Scalar beta, Scalar alpha) {
+  TORCH_CHECK(self.dim() == 2, "input must be a matrix");
+  TORCH_CHECK(m1.dim() == 2, "m1 must be a matrix");
+  TORCH_CHECK(m2.dim() == 2, "m2 must be a matrix");
+
+  TORCH_CHECK(
+      m1.size(1) == m2.size(0), "m1 and m2 shapes cannot be multiplied (",
+      m1.size(0), "x", m1.size(1), " and ", m2.size(0), "x", m2.size(1), ")");
+
+  TORCH_CHECK(
+      self.size(0) == m1.size(0) && self.size(1) == m2.size(1),
+      "input shape is incompatible with matrix multiplication (",
+      m1.size(0), "x", m1.size(1), " and ", m2.size(0), "x", m2.size(1), ")");
+
+  result.resize_as_(self);
```

Comment:
Do you know if this short circuits when `self` is the same as `result`?

Reply:
It short circuits if the sizes are the same, which would have the same effect:
https://github.com/pytorch/pytorch/blob/e41fe6086745387b3e1aa1ae6e05fe6037a06a57/aten/src/ATen/native/Resize.h#L36-L38